File: 그래프_위상정렬(topology-sort)/leetcode207(Medium)-Course Schedule.py
"""
이 문제는 <파이썬 알고리즘 인터뷰>에 나와있는 문제로,
책에서는 dfs 풀이로 나와있다. 하지만 이 문제는 매우 전형적인 위상정렬 문제.
"""

# 위상정렬 풀이 : 93ms, 15.4MB
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        graph = collections.defaultdict(list)
        for v, w in prerequisites:
            graph[v].append(w)
            indegree[w] += 1
        
        q = collections.deque()
        for v in range(numCourses):
            if indegree[v] == 0:
                q.append(v)

        count = 0
        while q:
            count += 1
            v = q.popleft()
            for w in graph[v]:
                indegree[w] -= 1
                if indegree[w] == 0:
                    q.append(w)
                
        
        return count == numCourses
    
# dfs : 92ms, 17.5MB
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = collections.defaultdict(list)
        for v, w in prerequisites:
            graph[v].append(w)

        traced = set()
        visited = set()

        def dfs(v):
            if v in traced:
                return False
            if v in visited:
                return True
            
            traced.add(v)
            for w in graph[v]:
                if not dfs(w):
                    return False
            traced.remove(v)
            visited.add(v)
            return True


        for v in range(numCourses):
            if not dfs(v):
                return False
        
        return True

File: 그래프_위상정렬/leetcode207(Medium)-Course Schedule.js
// 위상정렬 : 90ms, 46.3MB
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
  const indegree = new Array(numCourses).fill(0)
  const graph = prerequisites.reduce((obj, [v, w]) => {
    if (!obj[v]) obj[v] = []
    obj[v].push(w)
    indegree[w] += 1
    return obj
  }, {})

  const q = []
  for (let node = 0; node < numCourses; node += 1) {
    if (indegree[node] === 0) {
      q.push(node)
    }
  }

  let count = 0
  while (q.length > 0) {
    count += 1
    v = q.shift()
    if (!graph[v]) {
      continue
    }
    for (const w of graph[v]) {
      indegree[w] -= 1
      if (indegree[w] === 0) {
        q.push(w)
      }
    }
  }

  return count === numCourses
}

// dfs : 75ms, 46.6MB
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function (numCourses, prerequisites) {
  const graph = prerequisites.reduce((obj, [v, w]) => {
    if (!obj[v]) obj[v] = []
    obj[v].push(w)
    return obj
  }, {})

  const traced = new Set()
  const visited = new Set()

  const dfs = (v) => {
    if (traced.has(v)) {
      return false
    }
    if (visited.has(v)) {
      return true
    }
    traced.add(v)
    if (graph[v]) {
      for (const w of graph[v]) {
        if (!dfs(w)) {
          return false
        }
      }
    }
    traced.delete(v)
    visited.add(v)
    return true
  }

  for (const v of Object.keys(graph)) {
    if (!dfs(v)) {
      return false
    }
  }
  return true
}
